### PR TITLE
User endpoint discovery database speed up queries [6588]

### PR DESCRIPTION
**Pull request by MiguelBarro**
This pull request meets the Github issue #636 and the pull request #637.

The main goal is speed up the PDP database queries by replacing actual vector collections with hash tables keeping our current allocation-free philosophy intact.

### CODE REVIEW

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 1]
```diff
@@ -481,9 +483,9 @@ bool PDP::has_reader_proxy_data(const GUID_t& reader)
     {
         if (pit->m_guid.guidPrefix == reader.guidPrefix)
         {
-            for (ReaderProxyData* rit : pit->m_readers)
+            for (auto rp : *pit->m_readers)
             {
-                if (rit->guid() == reader)
+                if (rp.second->guid() == reader)
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDP.cpp]: We could use `pit->m_readers->find` here

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 2]
```diff
@@ -520,9 +520,9 @@ bool PDP::has_writer_proxy_data(const GUID_t& writer)
     {
         if (pit->m_guid.guidPrefix == writer.guidPrefix)
         {
-            for (WriterProxyData* wit : pit->m_writers)
+            for (auto wp : *pit->m_writers)
             {
-                if (wit->guid() == writer)
+                if (wp.second->guid() == writer)
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDP.cpp]: We could use `pit->m_writers->find` here

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp [hunk 3]
```diff
@@ -29,8 +29,11 @@
 #include <fastrtps/qos/QosPolicies.h>
 #include <fastrtps/utils/TimeConversion.h>
 
+#include <rtps/builtin/data/ProxyHashTables.hpp>
+
 #include <mutex>
 #include <chrono>
+#include <unordered_set>
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: Is this necessary?

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp [hunk 4]
```diff
@@ -92,15 +96,8 @@ ParticipantProxyData::~ParticipantProxyData()
...
-    {
-        delete it;
-    }
-
-    for (WriterProxyData* it : m_writers)
-    {
-        delete it;
-    }
+    delete m_readers;
+    delete m_writers;
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: Old behavior deleted the proxy objects held by the collections, new behavior does not do that. This implies a memory leak.

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp (hunk 1 above)
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp]: In fact, we can just do `return pit->m_readers->find(reader.entityId) != pit->m_readers->end();`

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp (hunk 2 above)
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp]: In fact, we can just do `return pit->m_writers->find(writer.entityId) != pit->m_writers->end();`

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 5]
```diff
@@ -597,24 +595,27 @@ bool PDP::removeWriterProxyData(const GUID_t& writer_guid)
...
         if (pit->m_guid.guidPrefix == writer_guid.guidPrefix)
         {
-            for (WriterProxyData* wit : pit->m_writers)
+            auto wit = pit->m_writers->find(writer_guid.entityId);
+
+            if (wit != pit->m_writers->end())
             {
-                if (wit->guid() == writer_guid)
+                WriterProxyData *pW = wit->second;
+                if (pW->guid() == writer_guid)
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDP.cpp]: This check is not necessary

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp (hunk 3 above)
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: nope

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp (hunk 4 above)
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/data/ParticipantProxyData.cpp]: Initially we didn't keep pointers :( later I forgot to mend this behavior.

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp (hunk 1 above)
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp]: agreed, we can speed up by using hash table capabilities:
```
 ProxyHashTable<ReaderProxyData> & readers = *pit->m_readers;
 return readers.find(reader.entityId) != readers.end();
```

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp (hunk 2 above)
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp]: agreed, we can speed up by using hash table capabilities:
```
 ProxyHashTable<WriterProxyData> & writers = *pit->m_writers;
 return writers.find(writer.entityId) != writers.end();
```

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp (hunk 5 above)
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp]: agreed, the key should match the proxy.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/744/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/630/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/740/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/204/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/748/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/633/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/744/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/207/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/750/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/635/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/746/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/209/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/754/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/639/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/750/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/213/)

**Pull request merged**